core/src/main/java/nl/nn/adapterframework/filesystem/ExchangeFileSystem.java

```diff
@@ -772,56 +802,81 @@ public Map<String, Object> getAdditionalAttachmentProperties(Attachment a) throw
+				log.debug("creating folder [" + resolver.getFolderName() + "] in mailbox ["+resolver.getMailbox()+"]");
+				createFolder(resolver);
+				findResults = exchangeService.findFolders(basefolderId, new SearchFilter.IsEqualTo(FolderSchema.DisplayName, resolver.getFolderName()), new FolderView(Integer.MAX_VALUE));
+			}
+			if (findResults.getTotalCount()==0) {
+				log.debug("folder [" + resolver.getFolderName() + "] not found in mailbox ["+resolver.getMailbox()+"]");
+				return null;
+			}
+			if (log.isDebugEnabled()) {
+				log.debug("amount of folders with name: " + resolver.getFolderName() + " = " + findResults.getTotalCount());
+				log.debug("found folder with name: " + findResults.getFolders().get(0).getDisplayName());
+			}
+			folderId = findResults.getFolders().get(0).getId();
+
+			cache.registerResolversFolder(resolver, folderId);
 		}
-		FolderId folderId = findResults.getFolders().get(0).getId();
+
 		return folderId;
 	}
 
-	@Override
-	public void createFolder(String folderName) throws FileSystemException {
-		ExchangeService exchangeService = getConnection();
+
+	private void createFolder(ExchangeFileSystemResolver resolver) throws FileSystemException {
+		ExchangeService exchangeService = getConnection(resolver.getMailbox());
 		boolean invalidateConnectionOnRelease = false;
 		try {
 			Folder folder = new Folder(exchangeService);
-			folder.setDisplayName(folderName);
-			folder.save(new FolderId(basefolderId.getUniqueId()));
+			folder.setDisplayName(resolver.getFolderName());
+			folder.save(cache.getBaseFolderId(resolver.getMailbox()));
+			cache.registerFolder(resolver.getMailbox(), folder);
 		} catch (Exception e) {
 			invalidateConnectionOnRelease = true;
-			throw new FileSystemException("cannot create folder ["+folderName+"]", e);
+			throw new FileSystemException("cannot create folder ["+resolver.getFolderName()+"] in mailbox ["+resolver.getMailbox()+"]", e);
 		} finally {
 			releaseConnection(exchangeService, invalidateConnectionOnRelease);
 		}
 	}
 
+	@Override
+	public void createFolder(String folderName) throws FileSystemException {
+		createFolder(getResolver(folderName));
+	}
+
 
 	@Override
 	public void removeFolder(String folderName, boolean removeNonEmptyFolder) throws FileSystemException {
-		ExchangeService exchangeService = getConnection();
+		ExchangeFileSystemResolver resolver = getResolver(folderName);
+		ExchangeService exchangeService = getConnection(resolver.getMailbox());
 		boolean invalidateConnectionOnRelease = false;
+
 		try {
-			FolderId folderId = getFolderIdByFolderName(exchangeService, folderName, false);
+			FolderId folderId = getFolderIdByFolderName(exchangeService, folderName, false, false);
 			Folder folder = Folder.bind(exchangeService, folderId);
 			if(removeNonEmptyFolder) {
 				folder.empty(DeleteMode.HardDelete, true);
 			}
 			folder.delete(DeleteMode.HardDelete);
+			cache.deregisterFolder(resolver.getMailbox(), folder);
```

Comment:
Op het moment dat er in een project meerdere EWS senders worden gebruikt en er wordt door de applicatie regelmatig folders aangemaakt of verwijderd, dan bestaat er een kans dat andere instanties van de cache achter lopen op de waarheid en mogelijk kunnen er fouten optreden.

Hiervoor zie ik verschillende oplossingsrichtingen:

- Cache middels attribuut kunnen in-uit schakelen en middels Javadoc en Ibisdoc duidelijk maken voor welke use-cases het gebruik van de cache handig kan zijn, wellicht zelfs default uitgeschakeld.

- Static, er zijn dan ook nog verschillende mogelijkheden om dit te realiseren. De cache zou in zijn geheel static kunnen zijn maar dat heeft wellicht ook weer nadelen. Daarnaast zou er een list van instances bijgehouden kunnen worden op static niveau en de deregisterFolder ook static aan te bieden. Deze methode gaat vervolgens de list met instances af om overal de cache bij te werken.

Hoe denken jullie over het mogelijke issue en de mogelijke oplossingen?

Reply:
Ik denk dat het risico dat je beschrijft reeel is, en dat het met name voor zal komen in de situaties dat je dacht profijt te hebben van de cache.
Het lijkt me verstandig om eerst een robuust werkende MultiMailbox versie van het ExchangeFileSystem zonder cache te maken. Die kan dan later eventueel uitgebreid worden met een cache, mocht daar behoefte aan zijn.

Voor nu lijkt me de beste aanpak:
- duidelijk krijgen of move en copy van de ene mailbox naar een andere kan, en zo ja, hoe dan.
- als het kan, dan kan ik me niet voorstellen dat het nodig is om altijd een user te impersonaten op een connectie; dat lijkt me dan alleen nodig voor 'zoeken' in de juiste basis mailbox
- als het niet kan, dan moeten move en copy niet toestaan dat er naar een andere mailbox gemoved en copied wordt, en kunnen ze eenvoudiger worden.